File: bindings/python/examples/how_tos/accounts_and_addresses/consolidate_outputs.py
import os

from dotenv import load_dotenv

from iota_sdk import ConsolidationParams, Utils, Wallet, WalletOptions, FeatureType

# In this example we will consolidate basic outputs from an account with only an AddressUnlockCondition by sending
# them to the same address again.

# This example uses secrets in environment variables for simplicity which
# should not be done in production.
load_dotenv()

if 'STRONGHOLD_PASSWORD' not in os.environ:
    raise Exception('.env STRONGHOLD_PASSWORD is undefined, see .env.example')

wallet = Wallet(WalletOptions(storage_path=os.environ.get('WALLET_DB_PATH')))
wallet.set_stronghold_password(os.environ['STRONGHOLD_PASSWORD'])

# Sync wallet to make sure account is updated with outputs from previous
# examples.
wallet.sync()
print('Account synced')

# List unspent outputs before consolidation.
# The output we created with example `request_funds` and the basic output from `mint` have only one
# unlock condition and it is an `AddressUnlockCondition`, and so they are valid for consolidation. They have the
# same `AddressUnlockCondition`(the address of the wallet), so they will be consolidated into one
# output.
outputs = wallet.unspent_outputs()
print('Outputs BEFORE consolidation:')

for i, output_data in enumerate(outputs):
    print(f'OUTPUT #{i}')
    print(
        f'- address: #{Utils.hex_to_bech32(output_data.address.pub_key_hash, "rms")}')
    print(f'- amount: #{output_data.output.amount}')

    native_tokens = [feature for feature in output_data.output.features if feature.type == FeatureType.NativeToken]
    first_native_token = next(iter(native_tokens), None)
    print(f'- native tokens: #{first_native_token}')

print('Sending consolidation transaction...')

# Consolidate unspent outputs and print the consolidation transaction ID
# Set `force` to true to force the consolidation even though the
# `output_threshold` isn't reached.
transaction = wallet.consolidate_outputs(ConsolidationParams(force=True))
print('Transaction sent: ', transaction.transaction_id)

# Wait for the consolidation transaction to get confirmed
block_id = wallet.reissue_transaction_until_included(
    transaction.transaction_id)

print(
    f'Transaction included: {os.environ["EXPLORER_URL"]}/block/{block_id}'
)

# Sync account
wallet.sync()
print('Account synced')

# Outputs after consolidation
outputs = wallet.unspent_outputs()
print('Outputs AFTER consolidation:')
for i, output_data in enumerate(outputs):
    print(f'OUTPUT #{i}')
    print(
        f'- address: #{Utils.hex_to_bech32(output_data.address.pub_key_hash, "rms")}')
    print(f'- amount: #{output_data.output.amount}')

    native_tokens = [feature for feature in output_data.output.features if feature.type == FeatureType.NativeToken]
    first_native_token = next(iter(native_tokens), None)
    print(f'- native tokens: #{first_native_token}')

File: bindings/nodejs/examples/how_tos/accounts_and_addresses/consolidate-outputs.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { CommonOutput, Utils, Wallet, initLogger } from '@iota/sdk';

// This example uses secrets in environment variables for simplicity which should not be done in production.
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./how_tos/accounts_and_addresses/consolidate-outputs.ts

// In this example we will consolidate basic outputs from an account with only an AddressUnlockCondition by sending
// them to the same address again.
async function run() {
    initLogger();
    try {
        for (const envVar of [
            'WALLET_DB_PATH',
            'STRONGHOLD_PASSWORD',
            'EXPLORER_URL',
        ])
            if (!(envVar in process.env)) {
                throw new Error(
                    `.env ${envVar} is undefined, see .env.example`,
                );
            }

        const wallet = new Wallet({
            storagePath: process.env.WALLET_DB_PATH,
        });

        const account = await wallet.getAccount('Alice');

        // To create an address we need to unlock stronghold.
        await wallet.setStrongholdPassword(
            process.env.STRONGHOLD_PASSWORD as string,
        );

        // Sync account to make sure account is updated with outputs from previous examples
        account.sync();
        console.log('Account synced');

        // List unspent outputs before consolidation.
        // The output we created with example `request_funds` and the basic output from `mint` have only one
        // unlock condition and it is an `AddressUnlockCondition`, and so they are valid for consolidation. They have the
        // same `AddressUnlockCondition`(the first address of the account), so they will be consolidated into one
        // output.
        const outputs = await account.unspentOutputs();
        console.log('Outputs BEFORE consolidation:');

        outputs.forEach(({ output, address }, i) => {
            console.log(`OUTPUT #${i}`);
            console.log(
                '- address: %s\n- amount: %d\n- native tokens: %s',
                Utils.hexToBech32(address.toString(), 'rms'),
                output.getAmount(),
                output instanceof CommonOutput
                    ? (output as CommonOutput).getNativeTokens()
                    : [],
            );
        });

        console.log('Sending consolidation transaction...');

        // Consolidate unspent outputs and print the consolidation transaction ID
        // Set `force` to true to force the consolidation even though the `output_threshold` isn't reached
        const transaction = await account.consolidateOutputs({
            force: true,
        });
        console.log('Transaction sent: %s', transaction.transactionId);

        // Wait for the consolidation transaction to get confirmed
        const blockId = account.retryTransactionUntilIncluded(
            transaction.transactionId,
        );

        console.log(
            'Transaction included: %s/block/$s',
            process.env.EXPLORER_URL,
            blockId,
        );

        // Sync account
        account.sync();
        console.log('Account synced');

        // Outputs after consolidation
        console.log('Outputs AFTER consolidation:');
        outputs.forEach(({ output, address }, i) => {
            console.log(`OUTPUT #${i}`);
            console.log(
                '- address: %s\n- amount: %d\n- native tokens: %s',
                Utils.hexToBech32(address.toString(), 'rms'),
                output.getAmount(),
                output instanceof CommonOutput
                    ? (output as CommonOutput).getNativeTokens()
                    : undefined,
            );
        });
    } catch (error) {
        console.error('Error: ', error);
    }
}

run().then(() => process.exit());
